controllers/operatorpolicy_status.go

```diff
@@ -0,0 +1,429 @@
+// preserving the `CreatedByPolicy` property on relatedObjects.
+//
+// This function assumes that all given related objects are of the same kind.
+//
+// Note that only changing the related objects will not emit a new compliance event, but will update
+// the status.
+func (r *OperatorPolicyReconciler) updateStatus(
+	ctx context.Context,
+	policy *policyv1beta1.OperatorPolicy,
+	updatedCondition metav1.Condition,
+	updatedRelatedObjs ...policyv1.RelatedObject,
+) error {
+	condChanged := false
+
+	if updatedCondition.LastTransitionTime.IsZero() {
+		updatedCondition.LastTransitionTime = metav1.Now()
+	}
+
+	condIdx, existingCondition := policy.Status.GetCondition(updatedCondition.Type)
+	if condIdx == -1 {
+		condChanged = true
+
+		// Just append, the conditions will be sorted later.
+		policy.Status.Conditions = append(policy.Status.Conditions, updatedCondition)
+	} else if conditionChanged(updatedCondition, existingCondition) {
+		condChanged = true
+
+		policy.Status.Conditions[condIdx] = updatedCondition
+	}
+
+	if condChanged {
+		updatedComplianceCondition := calculateComplianceCondition(policy)
+
+		compCondIdx, _ := policy.Status.GetCondition(updatedComplianceCondition.Type)
+		if compCondIdx == -1 {
+			policy.Status.Conditions = append(policy.Status.Conditions, updatedComplianceCondition)
+		} else {
+			policy.Status.Conditions[compCondIdx] = updatedComplianceCondition
+		}
+
+		// Sort the conditions based on their type.
+		sort.SliceStable(policy.Status.Conditions, func(i, j int) bool {
+			return policy.Status.Conditions[i].Type < policy.Status.Conditions[j].Type
+		})
+
+		if updatedComplianceCondition.Status == metav1.ConditionTrue {
+			policy.Status.ComplianceState = policyv1.Compliant
+		} else {
+			policy.Status.ComplianceState = policyv1.NonCompliant
+		}
+
+		err := r.emitComplianceEvent(ctx, policy, updatedComplianceCondition)
+		if err != nil {
+			return err
+		}
+	}
+
+	relObjsChanged := false
+
+	prevRelObjs := policy.Status.RelatedObjsOfKind(updatedRelatedObjs[0].Object.Kind)
+	if len(prevRelObjs) == len(updatedRelatedObjs) {
+		for _, prevObj := range prevRelObjs {
+			nameFound := false
+
+			for i, updatedObj := range updatedRelatedObjs {
+				if prevObj.Object.Metadata.Name == updatedObj.Object.Metadata.Name {
+					nameFound = true
+
+					if updatedObj.Properties != nil && prevObj.Properties != nil {
+						if updatedObj.Properties.UID != prevObj.Properties.UID {
+							relObjsChanged = true
+						} else if prevObj.Properties.CreatedByPolicy != nil {
+							// There is an assumption here that it will never need to transition to false.
+							updatedRelatedObjs[i].Properties.CreatedByPolicy = prevObj.Properties.CreatedByPolicy
+						}
+					}
+
+					if prevObj.Compliant != updatedObj.Compliant {
+						relObjsChanged = true
+					} else if prevObj.Reason != updatedObj.Reason {
+						relObjsChanged = true
+					}
+				}
+			}
+
+			if !nameFound {
+				relObjsChanged = true
+			}
```

Comment:
Is it safe to `break` if `relObjsChanged = true` here?

Reply:
Yes.